Implement UdpSocket for both Syn. & Async modes (Send & Receive functionality)

reza-ebrahimi opened the pull request:
This PR:
- Implements UDP send functionality with `boost::asio` library in `UdpSender` class.
- Refactors UDP receiver functionality to a new `UdpReceiver` class.

This PR closes issue #12.

reza-ebrahimi commented:
I will add `UdpSender` tests  to this PR by tomorrow.

reza-ebrahimi commented:
- Both **synchronous** and **asynchronous** send and receive functionality are supported.
- `UdpSocket` in a process using just one instance of `IoContext` object that can be used as single/multi thread IO context.
- **IoContext**
  - **UdpSocket**
  - **TcpSocket** (Implementable)
  - **Serial** (Implementable)

reza-ebrahimi commented:
- Both **synchronous** and **asynchronous** send and receive functionality are supported.
- Every `UdpSocket` in a process using just one instance of `IoContext` object that can be used as single/multi thread IO context.
- **IoContext**
  - **UdpSocket**
  - **TcpSocket** (Implementable)
  - **Serial** (Implementable)

reza-ebrahimi commented:
Current PR Status:

* Implemented `IoContext` (The main I/O Context that can be configured as single or multi thread with async task execution support)
* Implemented `UdpSocket` (The UDP socket functionality that working in blocking and non-blocking (Async with task execution) modes)
* Implemented `UdpDriver` Facade (Facade design pattern implementation that acts as a sender and receiver UDP socket pair container)
* Implemented `ROS2` and `raw` buffer message converters for Int, UInt and Float variants in std_msgs namespace.
* Implemented `UdpDriverNode` example and its test.
* Implemented IoContext, UdpSocket and UdpDriver tests.

reza-ebrahimi commented:
Current PR Status:

* Implemented `IoContext` (The main I/O Context that can be configured as single or multi thread with async task execution support)
* Implemented `UdpSocket` (The UDP socket functionality that working in **blocking** and **non-blocking** (Async with task execution) modes)
* Implemented `UdpDriver` Facade (Facade design pattern implementation that acts as a sender and receiver UDP socket pair container)
* Implemented `ROS2` and `raw` buffer message converters for **Int**, **UInt** and **Float** variants in **std_msgs** namespace.
* Implemented `UdpDriverNode` example and its test.
* Implemented IoContext, UdpSocket and UdpDriver tests.

flynneva commented:
@reza-ebrahimi is this desired to move `udp_driver` up into the parent `transport_drivers` directory?

reza-ebrahimi commented:
@flynneva Both serial port and UDP socket are using `boost::asio::io_service`, So potentially all IO operations (UDP, TCP and Serial) could be implemented by using `IoContext`.

This is my plan for project structure:
```javascript
- transport_drivers (root)
  - examples
    - UdpDriverNode
    - TcpDriverNode
    - SerialPortNode
  - include (header files)
    - Converters
      - StdMsgs
    - IoContext
    - UdpSocket
    - TcpSocket
    - SerialPortSocket
    - UdpDriver
    - TcpDriver
    - SerialPortDriver
  - src (source files)
    - Converters
      - StdMsgs
    - IoContext
    - UdpSocket
    - TcpSocket
    - SerialPortSocket
    - UdpDriver
    - TcpDriver
    - SerialPortDriver
  - test
    - IoContextTest
    - UdpSocketTest
    - TcpSocketTest
    - SerialPortTest
    - UdpDriverTest
    - TcpDriverTest
    - SerialPortDriverTest
    - UdpDriverNodeTest
    - TcpDriverNodeTest
    - SerialPortDriverNodeTest
packages.xml
CMakeLists.txt
```

reza-ebrahimi commented:
@flynneva Both serial port and UDP socket are using `boost::asio::io_service`, So potentially all IO operations (UDP, TCP and Serial) could be implemented by using `IoContext`.

This is my plan for project structure:

```
- transport_drivers (root)
    - examples
        - UdpDriverNode
        - TcpDriverNode
        - SerialPortNode
    - include (header files)
        - Converters
            - StdMsgs
        - IoContext (This instance is shared among all sockets in a single process)
        - UdpSocket (Using Shared IoContext)
        - TcpSocket (Using Shared IoContext)
        - SerialPortSocket (Using Shared IoContext)
        - UdpDriver (Facade for sender & receiver pair sockets)
        - TcpDriver (Facade for sender & receiver pair sockets)
        - SerialPortDriver (Facade for sender & receiver pair sockets)
    - src (source files)
        - Converters
            - StdMsgs
        - IoContext
        - UdpSocket 
        - TcpSocket
        - SerialPortSocket
        - UdpDriver
        - TcpDriver
        - SerialPortDriver
    - test
        - IoContextTest
        - UdpSocketTest
        - TcpSocketTest
        - SerialPortTest
        - UdpDriverTest
        - TcpDriverTest
        - SerialPortDriverTest
        - UdpDriverNodeTest
        - TcpDriverNodeTest
        - SerialPortDriverNodeTest
packages.xml
CMakeLists.txt
```

flynneva commented:
i know @JWhitleyWork wanted to still have the option to be able to inherit the `UdpDriver` class and make a node right off of that. I think you should still be able to do that using this directory structure though.....what do you think @JWhitleyWork?

JWhitleyWork commented on CHANGELOG.rst:
```diff
@@ -2,6 +2,17 @@
+* Implemented UdpDriverNode example and its test.
+* Implemented IoContext, UdpSocket and UdpDriver tests.
+* Contributors: Reza Ebrahimi
+
```
This is not the purpose of the CHANGELOG. Your changes do not constitute a release. Please remove these changes.

reza-ebrahimi replied:
What is the purpose of change log? Why I should remove contributors part from change log?

reza-ebrahimi replied:
If your concern is release version, We can change it. But I didn't understand why I should remove the implemented API log and contributors part.

reza-ebrahimi replied:
What is the purpose of change log? If your concern is release version, We can change it. But I didn't understand why I should remove the implemented API log and contributors part.

gavanderhoorn replied:
Changelogs are automatically generated as part of the release process. Same for version increments.

@JWhitleyWork is correct in that it's not necessary (nor perhaps desirable) to change this file by hand before running the tools.

The suggestion is actually to revert all your changes to the `CHANGELOG.rst` file.

reza-ebrahimi replied:
OK. I got it, will revert it.

reza-ebrahimi replied:
Done with 9a82e24abe471c61021e8cd86e3f734dbff0b123.

JWhitleyWork commented on README.md:
```diff
@@ -1,8 +1,17 @@
+    
+* **Serial Driver**
+    * Planning to implement with new `IoContext` (Blocking and NonBlocking) functionality.
+    * Current implementation contains legacy driver code (A package which contains a templated C++ class (which inherits from `rclcpp::Node`) which encapsulates basic receiving (and soon sending) of serial data.)
```
I'm not fond of deprecating the existing implementation with just an IOU for a new implementation. Please create classes with common functionality instead of moving the UDP version to the base. We want to keep the UDP driver and the Serial driver in separate packages.

reza-ebrahimi replied:
Since there is no provided reason behind two separate packages, I will move UDP to its package and leave serial package as before.

reza-ebrahimi replied:
No one deprecated Serial port. I was planned for implementing Serial port socket like the UDP one after finalizing UDP.
After that we can deprecate legacy code.

Since there is no provided reason behind two separate packages, But its OK, I will move UDP to its own package and leave serial package as before.

reza-ebrahimi replied:
Done 78c1d1c6925d6fbfa505c690b349c70f33e2b2ed.

JWhitleyWork reviewed (changes requested):
Added high-level concerns/problems with this PR.

flynneva commented:
@reza-ebrahimi i think you're desire to use a single `IoContext` for both the `serial_driver` and `udp_driver` is still possible. the `IoContext` has no ROS-specific dependencies and could be made into a pure  cpp shared library in a peer-directory to `serial_driver` and `udp_driver`, where both of those ROS packages then depend on that shared library.

JWhitleyWork commented:
@reza-ebrahimi There are *a lot* of linting and test errors on your branch. To view the errors, please run `colcon build --packages-select udp_driver serial_driver && colcon test --packages-select udp_driver serial_driver && colcon test-result --verbose`. These will need to be fixed before this PR can be merged.

flynneva commented:
@JWhitleyWork ive fixed most of them on my branch, i can make a PR from mine into @reza-ebrahimi to knock out two birds with one stone

JWhitleyWork commented:
@flynneva That would be awesome! Thanks! Mention me in it and I'll take a quick look as well, if you want.

reza-ebrahimi commented:
> @reza-ebrahimi There are _a lot_ of linting and test errors on your branch. To view the errors, please run `colcon build --packages-select udp_driver serial_driver && colcon test --packages-select udp_driver serial_driver && colcon test-result --verbose`. These will need to be fixed before this PR can be merged.

Is there any tool for these lint errors to do some of them automatically (Such as formatting)?

reza-ebrahimi commented:
> @reza-ebrahimi There are _a lot_ of linting and test errors on your branch. To view the errors, please run `colcon build --packages-select udp_driver serial_driver && colcon test --packages-select udp_driver serial_driver && colcon test-result --verbose`. These will need to be fixed before this PR can be merged.

@JWhitleyWork Is there any tool for these lint errors to do some of them automatically (Such as formatting)?

reza-ebrahimi commented:
@flynneva If you are interested, Please work on serial driver using this package `IoContext` class.

reza-ebrahimi commented:
> @flynneva That would be awesome! Thanks! Mention me in it and I'll take a quick look as well, if you want.

@JWhitleyWork Why you are encouraging to parallel work while this PR still open?

flynneva commented:
@reza-ebrahimi i just made a PR to your 'main' branch. If you merge it, it should fix the linter errors for this PR.

reza-ebrahimi commented:
@flynneva PRs are just for pushing changes between branches by PR creator, They are not a tool for collaboration between different people, This is not a best practice.

Besides if you are trying to push your changes to this PR, that's Ok from my side.

flynneva commented:
@reza-ebrahimi I think we're all on the same page here. I've made a PR to your main branch, I couldn't get the io_context lib to link correctly so maybe you or @JWhitleyWork could help me out there. that pr also fixes the lint errors

reza-ebrahimi commented:
Since this PR is just for udp_driver please just push your linter fixes in a different PR. I'm recommending create another PR for your changes on serial driver.

flynneva commented:
@reza-ebrahimi  to be fair you just mentioned above that you wanted to use the `io_context` for both the `serial_driver` and for `udp_driver`. if this PR still has that goal, you'll need to move the `io_context` to its own shared library so that a future PR can add this to the `serial_driver`.

[the PR I just made](https://github.com/reza-ebrahimi/transport_drivers/pull/1) not only fixes the linter errors but also begins to move the `io_context` to its own shared lib.

if you'd like i can remove the `udp_component` bit I added, but really i think it should also be included in this PR. let me know and I can go from there.

reza-ebrahimi commented:
@flynneva Please just add lint errors fixes to a different PR. Your changes contains entities like udp_component that doesn't related to current design in this PR.

flynneva commented:
there are other minor changes (like moving `udp_driver` to a sub-directory within the `include` directory) to follow the standards found here in the docs https://docs.ros.org/en/foxy/Tutorials/Ament-CMake-Documentation.html#id3

JWhitleyWork commented:
@reza-ebrahimi You're being somewhat difficult about this. @flynneva is trying to help by making your branch work correctly and fixing the problems that you either haven't or won't. It's fine for him to remove the additional code for `udp_component` because I agree that is orthogonal to this PR but the linter and formatting fixes are required for this PR to be merged.

Once @flynneva completes his PR to your branch, please review his changes. If they are not acceptable, you at least need to fix the linting and formatting errors in this PR.

flynneva commented:
just made a new one with `udp_component` removed. i still left the `io_context` in its own directory though since i think that should be included in this PR.

https://github.com/reza-ebrahimi/transport_drivers/pull/2

reza-ebrahimi commented:
> @reza-ebrahimi  to be fair you just mentioned above that you wanted to use the `io_context` for both the `serial_driver` and for `udp_driver`. if this PR still has that goal, you'll need to move the `io_context` to its own shared library so that a future PR can add this to the `serial_driver`.
> 
> [the PR I just made](https://github.com/reza-ebrahimi/transport_drivers/pull/1) not only fixes the linter errors but also begins to move the `io_context` to its own shared lib. this PR does not make any changes yet to the `serial_driver`.
> 
> if you'd like i can remove the `udp_component` bit I added, but really i think it should also be included in this PR. let me know and I can go from there.

You are right. But this could happen when all drivers take their places in just one package, I reverted that structure that you know. 

For now we have two drivers:
- udp_driver
- serial_driver

If we want to share IoContext in both drivers, we should share it in root of the repository, it could be a static or dynamic library, or just both drivers reference to it's files statically using linking to object files (normal include).

I will help you to implement serial_driver based on this design if you want.

reza-ebrahimi commented:
> @reza-ebrahimi You're being somewhat difficult about this. @flynneva is trying to help by making your branch work correctly and fixing the problems that you either haven't or won't. It's fine for him to remove the additional code for `udp_component` because I agree that is orthogonal to this PR but the linter and formatting fixes are required for this PR to be merged.
> 
> Once @flynneva completes his PR to your branch, please review his changes. If they are not acceptable, you at least need to fix the linting and formatting errors in this PR.

@JWhitleyWork I'm appreciate @flynneva helps, There is no problem. All I'm trying to do is to never mistaken a PR with a repository.

I will review @flynneva changes.

flynneva commented:
>If we want to share IoContext in both drivers, we should share it in root of the repository, it could be a static or dynamic library, or just both drivers reference to it's files statically using linking to object files (normal include).

@reza-ebrahimi so thats what i was trying to do and I think have 90% there....just not linking correctly. maybe you can figure out whats wrong in my PR

reza-ebrahimi commented:
@JWhitleyWork Please take a review, All linter errors are fixed.

```bash
$ colcon build --packages-select udp_driver && colcon test --packages-select udp_driver && colcon test-result --verbose
Starting >>> udp_driver
Finished <<< udp_driver [0.33s]                     

Summary: 1 package finished [0.43s]
Starting >>> udp_driver
Finished <<< udp_driver [6.07s]          

Summary: 1 package finished [6.21s]
Summary: 100 tests, 0 errors, 0 failures, 0 skipped
```

JWhitleyWork commented:
@reza-ebrahimi Thanks for working with @flynneva to get these resolved! However, it looks like one of your branches may have been out-of-date when you started working on it as there are now conflicts that must be resolved before I can merge. See the warning below.

reza-ebrahimi commented:
Thanks @JWhitleyWork, let me check it.

reza-ebrahimi commented:
@JWhitleyWork I have just one `main` branch. I see these conflicts here in PR, but I don't know where they are happening.

JWhitleyWork commented:
@reza-ebrahimi I'll take a look today.

JWhitleyWork commented:
@reza-ebrahimi So here's the root of the problem: You made your changes based on a very old version of the `main` branch of this repository - back from version 0.0.2. @flynneva then made his changes based on your changes. If I revert the PR that @flynneva made and rebase your branch on the current `main`, your changes are all fine. However, since @flynneva's changes were based on your changes, which were based on an old version of `main`, merging his changes causes conflicts. In order to solve this, I have reverted the merge request from @flynneva on your branch. I also reverted the name change from `udp_driver` to `transport_driver` (we want to maintain both separate packages - `udp_driver` and `serial_driver`). I will now apply the linter fixes to your branch as well and you can review the PR when I'm done to make sure that your changes are as you expect.

reza-ebrahimi commented:
@JWhitleyWork You can simply copy and paste my last changes to the PR you want to make.
@flynneva Changes are older than my last changes, I have made lots of changes on top of them.

reza-ebrahimi commented:
@JWhitleyWork I applied linter fixes again. Please review it.

esteve commented:
@reza-ebrahimi @JWhitleyWork @flynneva jumping a little late to the party, but I have two questions:

- Why use Boost.Asio instead of the non-Boost version of Asio?
- In https://github.com/ros-drivers/transport_drivers/pull/31/files#diff-82eba09261662aabb4bbab508fe1f6306f6bd9daaf022774149c6f627f33d6d4R57-R58, it seems to me that only `Int32` data can be processed. Is that correct? Does the API accept other types of messages?

flynneva commented:
@esteve switching to just the plain Asio would make a lot of sense moving forward. Make a PR? And yes I'm working on switching to a more generic receiver and sender node this week

esteve commented:
@flynneva I started working on a Boost-less version right after sending hitting send :sweat_smile: I've submitted a draft PR https://github.com/ros-drivers/transport_drivers/pull/39, the only missing part is reimplementing `boost::thread_group` in C++11/14/17

reza-ebrahimi commented:
@flynneva The current design supports every message type, Just implementing `From` & `To` converters is enough.

For `sensor_msgs::NavSatFix` message type:

```bash
namespace drivers
{
namespace common
{

void convertToRos2Message(const MutSocketBuffer & in, sensor_msgs::NavSatFix & out)
{
}

void convertFromRos2Message(const sensor_msgs::NavSatFix::SharedPtr & in, MutSocketBuffer & out)
{
}

}  // namespace common
}  // namespace drivers
```

JWhitleyWork commented:
Much appreciated, gentlemen! Yes, I agree that the stand-alone ASIO is a better choice.
